doc/developer-guide/io-framework.md

```diff
@@ -0,0 +1,219 @@
+argument that will be passed to each callback.
+
+The returned value can be a negative error code if there has been any wait
+problem while initializing the system, or a positive error number if the
+system has been correctly initialized but it didn't complete cleany. If
+everything finished fine, 0 will be returned.
+
+### Termination
+
+When it's determined that the process must be terminated, a call to
+`gf_io_shutdown()` must be done.
+
+```c
+void gf_io_shutdown(void);
+```
+
+This function initiates a shutdown procedure, but returns immediately.
+Once the shutdown is completed, `gf_io_run()` will return. It can be
+called from anywhere.
+
+When shutdown is initiated, all I/O should have been stopped. If there
+is active I/O during the shutdown, they can complete, fail or be cancelled,
+depending on what state the request was. To ensure consistent behavior, try
+to always stop I/O before terminating the I/O framework.
+
+### Normal operation
+
+After everything is ready, the normal operation of the I/O framework is
+very simple:
+
+1. A worker peeks one completion event from the kernel.
+2. The callback associated to the completion event is executed.
+   2.1. The callback can prepare new I/O requests using one of the
+        `gf_io_*` I/O functions available for I/O operations.
+   2.2. Requests are added to a private queue in the current worker
+        using `gf_io_worker_add()`.
+   2.3. The callback may (but it's not required to) explicitly flush all
+        queued requests to the kernel using `gf_io_worker_flush()`.
+3. Once the callback finishes, any queued requests are automatically
+   flushed.
+
+### Available I/O operations
+
+The `gf_io_request_t` object used to manage the request is recommended to
+be embedded into bigger context structures so that a request can easily
+be mapped to its containing object using `gf_io_object()`. This way it's
+very easy to pass all needed information between the caller and the callback
+without additional memory allocations.
+
+All operations will also have a `data` argument to pass any additional
+per-request private data that the callback may need. This data will be
+available in `req->data`.
+
+Many of the I/O operations will have a timeout argument, which represents
+the maximum time allowed for the I/O to complete. If the operation takes
+more than that time, the system call will be cancelled and the callback
+will be executed passing a `-ETIMEDOUT` error.
+
+I/O operations will also have a priority argument that makes it possible
+to give different priorities to each requests so that the kernel scheduler
+can efficiently manage them based on their priority.
+
+#### Asynchronous call
+
+```c
+void gf_io_async(gf_io_request_t *req, gf_io_callback_t *callback,
+                 void *data);
+```
+
+#### Read operation
+
+```c
+void gf_io_readv(gf_io_request_t *req, gf_io_callback_t *callback, void *data,
+                 int32_t fd, uint64_t offset, struct iovec *iovec,
```

Comment:
nit: would be great to align variable order as preadv. Same below for the writev counterpart.

Reply:
sure